pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+//  backed off retries more details on inline code comments in the func
+
+// syncOne reconciles a CronJob with a list of any Jobs that it created.
+// All known jobs created by "cj" should be included in "js".
+// The current time is passed in to facilitate testing.
+// It has no receiver, to facilitate testing.
+func syncOne2(
+	cj *batchv1beta1.CronJob,
+	js []batchv1.Job,
+	now time.Time,
+	jc jobControlInterface,
+	cjc cjControlInterface,
+	recorder record.EventRecorder) (*time.Duration, error) {
+
+	childrenJobs := make(map[types.UID]bool)
+	for _, j := range js {
+		childrenJobs[j.ObjectMeta.UID] = true
+		found := inActiveList(*cj, j.ObjectMeta.UID)
+		if !found && !IsJobFinished(&j) {
+			recorder.Eventf(cj, corev1.EventTypeWarning, "UnexpectedJob", "Saw a job that the controller did not create or forgot: %s", j.Name)
+			// We found an unfinished job that has us as the parent, but it is not in our Active list.
+			// This could happen if we crashed right after creating the Job and before updating the status,
+			// or if our jobs list is newer than our cj status after a relist, or if someone intentionally created
+			// a job that they wanted us to adopt.
+
+			// TODO: maybe handle the adoption case?  Concurrency/suspend rules will not apply in that case, obviously, since we can't
+			// stop users from creating jobs if they have permission.  It is assumed that if a
+			// user has permission to create a job within a namespace, then they have permission to make any cronJob
+			// in the same namespace "adopt" that job.  ReplicaSets and their Pods work the same way.
+			// TBS: how to update cj.Status.LastScheduleTime if the adopted job is newer than any we knew about?
+		} else if found && IsJobFinished(&j) {
+			_, status := getFinishedStatus(&j)
+			deleteFromActiveList(cj, j.ObjectMeta.UID)
+			recorder.Eventf(cj, corev1.EventTypeNormal, "SawCompletedJob", "Saw completed job: %s, status: %v", j.Name, status)
+		}
+	}
+
+	// Remove any job reference from the active list if the corresponding job does not exist any more.
+	// Otherwise, the cronjob may be stuck in active mode forever even though there is no matching
+	// job running.
+	for _, j := range cj.Status.Active {
+		if found := childrenJobs[j.UID]; !found {
+			recorder.Eventf(cj, corev1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
+			deleteFromActiveList(cj, j.UID)
+		}
+	}
+
+	// TODO: @alpatel explore if cached client can be used as realCJControl
+	updatedCJ, err := cjc.UpdateStatus(cj)
+	if err != nil {
+		klog.V(2).InfoS("Unable to update status for cronjon", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "resourceVersion", cj.ResourceVersion, "err", err)
+		return nil, err
+	}
+	*cj = *updatedCJ
+
+	if cj.DeletionTimestamp != nil {
+		// The CronJob is being deleted.
+		// Don't do anything other than updating status.
+		return nil, fmt.Errorf("cronjob %s/%s is being deleted", cj.Namespace, cj.Name)
```

Comment:
Why do you return an error?  I would expect errors to end up retried, but if the deletiontimestamp is set, you don't need to requeue for an error..  You just need to update status and return normally, right?

Reply:
already updating status in the previous block, can return without error.